cmd/cluster-image-registry-operator/main.go

```diff
@@ -24,59 +28,93 @@ import (
 	klog.Infof("Go OS/Arch: %s/%s", runtime.GOOS, runtime.GOARCH)
 }
 
 func main() {
-	cmd := &cobra.Command{
-		Use:   "cluster-image-registry-operator",
-		Short: "OpenShift cluster image registry operator",
-		Run:   runOperator,
-	}
-	cmd.AddCommand(watchdog.NewFileWatcherWatchdog())
-	if err := cmd.Execute(); err != nil {
-		klog.Errorf("%v", err)
-		os.Exit(1)
-	}
-}
-
-// runOperator starts image registry operator and is our default command when
-// no other parameter is provided on the command line.
-func runOperator(cmd *cobra.Command, args []string) {
 	klogFlags := flag.NewFlagSet("klog", flag.ExitOnError)
 	klog.InitFlags(klogFlags)
-
-	logstderr := klogFlags.Lookup("logtostderr")
-	if logstderr != nil {
+	if logstderr := klogFlags.Lookup("logtostderr"); logstderr != nil {
 		_ = logstderr.Value.Set("true")
 	}
 
-	printVersion()
-
-	rand.Seed(time.Now().UnixNano())
-
-	cfg, err := regopclient.GetConfig()
-	if err != nil {
-		klog.Fatalf("Error building kubeconfig: %s", err)
-	}
-
-	// set up signals so we handle the first shutdown signal gracefully
-	stopCh := signals.SetupSignalHandler()
-
-	go metrics.RunServer(metricsPort, stopCh)
-
+	watchedFileChanged := make(chan struct{})
 	ctx, cancel := context.WithCancel(context.Background())
+	stopCh := signals.SetupSignalHandler()
 	go func() {
 		defer cancel()
-		<-stopCh
-		klog.Infof("Received SIGTERM or SIGINT signal, shutting down the operator.")
+		select {
+		case <-stopCh:
+			klog.Infof("Received SIGTERM or SIGINT signal, shutting down the operator.")
+		case <-watchedFileChanged:
+			klog.Infof("Watched file changed, shutting down the operator.")
+		}
 	}()
 
-	err = operator.RunOperator(ctx, cfg)
-	if err != nil {
-		klog.Fatal(err)
+	cmd := &cobra.Command{
+		Use:   "cluster-image-registry-operator",
+		Short: "OpenShift cluster image registry operator",
+		Run: func(cmd *cobra.Command, args []string) {
+			ctrl := controllercmd.NewController(
+				"image-registry-operator",
+				func(ctx context.Context, cctx *controllercmd.ControllerContext) error {
+					printVersion()
+					rand.Seed(time.Now().UnixNano())
+					go metrics.RunServer(metricsPort)
+					return operator.RunOperator(ctx, cctx.KubeConfig)
+				},
+			).WithLeaderElection(
+				configv1.LeaderElection{
+					LeaseDuration: metav1.Duration{
+						Duration: 15 * time.Second,
+					},
+					RenewDeadline: metav1.Duration{
+						Duration: 10 * time.Second,
+					},
+					RetryPeriod: metav1.Duration{
+						Duration: 2 * time.Second,
+					},
+				},
+				defaults.ImageRegistryOperatorNamespace,
+				defaults.LeaderLockConfigMapName,
+			)
+
+			if len(filesToWatch) > 0 {
+				fileContents := make(map[string][]byte)
+				watchPaths := make([]string, 0)
+				for _, path := range filesToWatch {
+					fcontent, err := ioutil.ReadFile(path)
+					if err != nil {
+						klog.Warningf("Unable to watch %s: %v", path, err)
+						continue
+					}
+					fileContents[path] = fcontent
```

Comment:
You don't need to provide this map unless you want to avoid race conditions. But to avoid race conditions, the file consumer should provide the value that was read from the file.

If the consumer cannot provide it, these is no reason to read it manually.

Reply:
This was rather confusing but yeah, seems like if we don't pass it it calculates on its own. Patching it.